eclair-core/src/main/scala/fr/acinq/eclair/payment/MultiPartPaymentHandler.scala

```diff
@@ -0,0 +1,143 @@
+/*
+ * Copyright 2019 ACINQ SAS
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package fr.acinq.eclair.payment
+
+import akka.actor.{ActorRef, Props}
+import fr.acinq.bitcoin.ByteVector32
+import fr.acinq.eclair.payment.PaymentReceived.PartialPayment
+import fr.acinq.eclair.wire.{FailureMessage, IncorrectOrUnknownPaymentDetails, UpdateAddHtlc}
+import fr.acinq.eclair.{FSMDiagnosticActorLogging, LongToBtcAmount, MilliSatoshi, NodeParams, wire}
+
+/**
+ * Created by t-bast on 18/07/2019.
+ */
+
+/**
+ * Handler for a multi-part payment (see https://github.com/lightningnetwork/lightning-rfc/blob/master/04-onion-routing.md#basic-multi-part-payments).
+ * Once all the partial payments are received, a MultiPartHtlcSucceeded message is sent to the parent.
+ * After a reasonable delay, if not enough partial payments have been received, a MultiPartHtlcFailed message is sent to the parent.
+ * This handler assumes that the parent only sends payments for the same payment hash.
+ */
+class MultiPartPaymentHandler(nodeParams: NodeParams, paymentHash: ByteVector32, totalAmount: MilliSatoshi, parent: ActorRef) extends FSMDiagnosticActorLogging[MultiPartPaymentHandler.State, MultiPartPaymentHandler.Data] {
+
+  import MultiPartPaymentHandler._
+
+  setTimer(PaymentTimeout.toString, PaymentTimeout, nodeParams.multiPartPaymentExpiry, repeat = false)
+
+  startWith(WAITING_FOR_HTLC, WaitingForHtlc(0 msat, Nil))
+
+  when(WAITING_FOR_HTLC) {
+    case Event(PaymentTimeout, d: WaitingForHtlc) =>
+      goto(PAYMENT_FAILED) using PaymentFailed(d.paidAmount, wire.PaymentTimeout, d.parts)
+
+    case Event(MultiPartHtlc(totalAmount2, htlc), d: WaitingForHtlc) =>
+      require(htlc.paymentHash == paymentHash, s"invalid payment hash (expected $paymentHash, received ${htlc.paymentHash}")
+      val pp = PartialPayment(htlc.amountMsat, htlc.channelId)
+      if (totalAmount != totalAmount2) {
+        log.warning(s"multi-part payment total amount mismatch: previously $totalAmount, now $totalAmount2")
+        goto(PAYMENT_FAILED) using PaymentFailed(d.paidAmount, IncorrectOrUnknownPaymentDetails(totalAmount2, nodeParams.currentBlockHeight), PendingPayment(htlc.id, pp, sender) :: d.parts)
```

Comment:
I would have used `+:` here, probably doesn't matter much.

```suggestion
        goto(PAYMENT_FAILED) using PaymentFailed(d.paidAmount, IncorrectOrUnknownPaymentDetails(totalAmount2, nodeParams.currentBlockHeight), PendingPayment(htlc.id, pp, sender) +: d.parts)
```

Reply:
A rewrite with less repetition:

```scala
case Event(mph@MultiPartHtlc(_, htlc), d: WaitingForHtlc) =>
  require(htlc.paymentHash == paymentHash, s"invalid payment hash (expected $paymentHash, received ${htlc.paymentHash}")
  val paidAmount1 = d.paidAmount + htlc.amountMsat
  val part = PendingPayment(htlc.id, PartialPayment(htlc.amountMsat, htlc.channelId), sender)
  val parts1 = part +: d.parts
  if (totalAmount != mph.totalAmount) {
    log.warning(s"multi-part payment total amount mismatch: previously $totalAmount, now ${mph.totalAmount}")
    goto(PAYMENT_FAILED) using PaymentFailed(d.paidAmount, IncorrectOrUnknownPaymentDetails(mph.totalAmount, nodeParams.currentBlockHeight), parts1)
  } else if (paidAmount1 >= totalAmount) {
    goto(PAYMENT_SUCCEEDED) using PaymentSucceeded(paidAmount1, parts1)
  } else {
    stay using d.copy(paidAmount = paidAmount1, parts = parts1)
  }
```

Reply:
NB: I think in the failure case we should arguably return `paidAmount1`:

```scala
goto(PAYMENT_FAILED) using PaymentFailed(paidAmount1, IncorrectOrUnknownPaymentDetails(mph.totalAmount, nodeParams.currentBlockHeight), parts1)

```